keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,418 @@
+to terminate a started job, which results in finishing the associated workload.
+
+Still the mitigation might be problematic as the timeout needs to include both
+the scheduling and execution time for a job, thus needs to be higher than one
+would want.
+
+Thus, we consider introduction of timeout on reaching the `PodsScheduled`
+condition by a workload since its job start (see:
+[Timeout on reaching the PodsScheduled condition](#timeout-on-reaching-the-podsscheduled-condition))
+as a possible future improvement.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### PodsScheduled workload condition
+
+First, we introduce a new workload condition, called `PodsScheduled`, to indicate
+that the first batch of workload's pods is scheduled, more precisely:
+`job.status.ready + job.status.succeeded >= job.spec.parallelism`.
+The condition is added by the kueue's Job controller when it reconciles a job on
+a Job status update. Note that, verifying the condition does not require extra
+API calls to the job or individual pods.
+
+### Workload API
+
+We extend the Workload API with the new `avoidConcurrentPodScheduling` field to
+enable the new behavior.
+
+```golang
+// WorkloadSpec defines the desired state of Workload
+type WorkloadSpec struct {
+  ...
+  // avoidConcurrentPodScheduling, when true, indicates that the workload blocks
+  // admission of other workloads in the cluster until it is in the
+  // `PodsScheduled` condition. Further, it delays admission of such workloads
+  // until all other admitted workloads are in the `PodsScheduled` condition.
+  // If false, the workload starts as soon as it is admitted and it does not
+  // block admission of other workloads. The `null` value results in the same
+  // behavior as `false`.
+	avoidConcurrentPodScheduling *bool `json:"avoidConcurrentPodScheduling,omitempty"`
+}
+```
+
+### Job annotation
+
+We introduce a new Job annotation, called `kueue.x-k8s.io/avoid-concurrent-pod-scheduling`,
+as a JobAPI counterpart of the `avoidConcurrentPodScheduling` Workload API field. When
+the annotation is present, the field value is set as `true` during workload
+creation.
+
+### The mechanism
+
+When a workload with `avoidConcurrentPodScheduling=true` is nominated for admission,
+kueue scheduler blocks its admission (and of other workloads nominated in
```

Comment:
for the cluster, or the ClusterQueue?

Reply:
for the entire cluster - added a note to clarify. We care about the underlying cluster nodes.